crates/ruma-common/src/push/condition/flattened_json.rs

```diff
@@ -0,0 +1,379 @@
+            Self::Integer(n) => n.serialize(serializer),
+            Self::String(s) => serializer.serialize_str(s),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for NonCompoundJsonValue {
+    #[inline]
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let val = JsonValue::deserialize(deserializer)?;
+        val.try_into().map_err(serde::de::Error::custom)
+    }
+}
+
+impl From<bool> for NonCompoundJsonValue {
+    fn from(value: bool) -> Self {
+        Self::Bool(value)
+    }
+}
+
+impl From<Int> for NonCompoundJsonValue {
+    fn from(value: Int) -> Self {
+        Self::Integer(value)
+    }
+}
+
+impl From<String> for NonCompoundJsonValue {
+    fn from(value: String) -> Self {
+        Self::String(value)
+    }
+}
+
+impl From<&str> for NonCompoundJsonValue {
+    fn from(value: &str) -> Self {
+        value.to_owned().into()
+    }
+}
+
+impl PartialEq<FlattenedJsonValue> for NonCompoundJsonValue {
+    fn eq(&self, other: &FlattenedJsonValue) -> bool {
+        match self {
+            Self::Null => *other == FlattenedJsonValue::Null,
+            Self::Bool(b) => other.as_bool().map_or(false, |other_b| *b == other_b),
+            Self::Integer(i) => other.as_integer().map_or(false, |other_i| *i == other_i),
+            Self::String(s) => other.as_str().map_or(false, |other_s| s == other_s),
+        }
+    }
+}
+
+/// Possible JSON values after an object is flattened.
+#[derive(Debug, Clone, Default, Eq, PartialEq)]
+#[allow(clippy::exhaustive_enums)]
+pub enum FlattenedJsonValue {
+    /// Represents a `null` value.
+    #[default]
+    Null,
+
+    /// Represents a boolean.
+    Bool(bool),
+
+    /// Represents an integer.
+    Integer(Int),
+
+    /// Represents a string.
+    String(String),
+
+    /// Represents an array.
+    Array(Vec<NonCompoundJsonValue>),
+}
+
+impl FlattenedJsonValue {
+    /// If the `FlattenedJsonValue` is a `Bool`, return the inner value.
+    pub fn as_bool(&self) -> Option<bool> {
+        match self {
+            Self::Bool(b) => Some(*b),
+            _ => None,
+        }
+    }
+
+    /// If the `FlattenedJsonValue` is an `Integer`, return the inner value.
+    pub fn as_integer(&self) -> Option<Int> {
+        match self {
+            Self::Integer(i) => Some(*i),
+            _ => None,
+        }
+    }
+
+    /// If the `FlattenedJsonValue` is a `String`, return a reference to the inner value.
+    pub fn as_str(&self) -> Option<&str> {
+        match self {
+            Self::String(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// If the `FlattenedJsonValue` is an `Array`, return a reference to the inner value.
+    pub fn as_array(&self) -> Option<&[NonCompoundJsonValue]> {
+        match self {
+            Self::Array(a) => Some(a),
+            _ => None,
+        }
+    }
+}
+
+impl TryFrom<JsonValue> for FlattenedJsonValue {
```

Comment:
Can we make this a private constructor function instead?

Reply:
For `ScalarJsonValue` too?

Reply:
Probably makes sense? This one has a comment about how it is used, didn't check the other one.

Reply:
Ah right, there's no comment in the other one, but I don't think the conversion would be useful outside of this crate.